matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/send/LocalEchoEventFactory.kt

```diff
@@ -158,4 +213,72 @@ internal class LocalEchoEventFactory(private val credentials: Credentials) {
     private fun dummyEventId(roomId: String): String {
         return roomId + "-" + dummyOriginServerTs()
     }
+
+    fun createReplyTextEvent(roomId: String, eventReplied: Event, replyText: String): Event? {
+        //Fallbacks and event representation
+        //TODO Add error/warning logs when any of this is null
+        val permalink = PermalinkFactory.createPermalink(eventReplied) ?: return null
+        val userId = eventReplied.sender ?: return null
+        val userLink = PermalinkFactory.createPermalink(userId) ?: return null
+//        <mx-reply>
+//            <blockquote>
+//                <a href="https://matrix.to/#/!somewhere:domain.com/$event:domain.com">In reply to</a>
+//                <a href="https://matrix.to/#/@alice:example.org">@alice:example.org</a>
+//                <br />
+//                <!-- This is where the related event's HTML would be. -->
+//            </blockquote>
+//        </mx-reply>
+//        This is where the reply goes.
+        val body = bodyForReply(eventReplied.content.toModel<MessageContent>())
+        val replyFallbackTemplateFormatted = """
+           <mx-reply><blockquote><a href="%s">${stringProvider.getString(R.string.in_reply_to)}</a><a href="%s">%s</a><br />%s</blockquote></mx-reply>%s
+        """.trimIndent().format(permalink, userLink, userId, body.second ?: body.first, replyText)
+//
+//        > <@alice:example.org> This is the original body
+//
+//        This is where the reply goes
+        val lines = body.first.split("\n")
+        val plainTextBody = StringBuffer("><${userId}>")
+        lines.firstOrNull()?.also { plainTextBody.append(" $it") }
+        lines.forEachIndexed { index, s ->
+            if (index > 0) {
+                plainTextBody.append("\n>$s")
+            }
+        }
+        plainTextBody.append("\n\n").append(replyText)
+
+        val eventId = eventReplied.eventId ?: return null
+        val content = MessageTextContent(
+                type = MessageType.MSGTYPE_TEXT,
+                format = MessageType.FORMAT_MATRIX_HTML,
+                body = plainTextBody.toString(),
+                formattedBody = replyFallbackTemplateFormatted,
+                relatesTo = RelationDefaultContent(null, null, ReplyToContent(eventId))
+        )
+        return createEvent(roomId, content)
+    }
+
+    private fun bodyForReply(content: MessageContent?): Pair<String, String?> {
```

Comment:
Create data class with named members, for code clarity, instead of using Pair?

Reply:
It's a private method of internal class, will add comments